### ISSUE TITLE
[rllib] ImpALA agent implementation is unstable

### ISSUE THREAD

**Issue by praveen-palanisamy** [ray-project/ray#2857]
### Describe the problem
The IMPALA agent implementation does not seem to be stable and in fact almost settles to a random policy if let to run for more number of steps.

The episode reward (mean) of IMPALA agent on Pong (PongNoFrameskip-v4) with 32 workers 
is shown below:
Run 1:
![pong-impala_impala_pongnoframeskip-v4](https://user-images.githubusercontent.com/4770482/45388765-fbb7ce00-b5e7-11e8-9f1a-1d59ea557437.PNG)

The issue is consistently reproducible and also exist in the case when vectorized env instances are used.
The episode reward (mean) on Pong (PongNoFrameskip-v4) with 32 workers, 10 env instance per worker (2 runs):

Run 2: 
![pong-impala-vectorized_impala_pongnoframeskip-v4_2](https://user-images.githubusercontent.com/4770482/45388699-babfb980-b5e7-11e8-9946-46332b3ebdf0.PNG)

Run 3:
![pong-impala-vectorized_impala_pongnoframeskip-v4_1](https://user-images.githubusercontent.com/4770482/45388577-63214e00-b5e7-11e8-94c2-6124ac425a44.PNG)

All the runs above used the same train batch size of 500 and 50 samples from each env instance and other parameters at their default values.

Combined graph from all runs:

![pong-impala_aio](https://user-images.githubusercontent.com/4770482/45389082-e42d1500-b5e8-11e8-8197-26edbeb3ab17.PNG)

Wonder what is causing the big, bad and sudden policy gradient update. It almost looks like a unrecoverable reset of the policy. The V-trace off-policy corrections were set to true in all of these runs and there was no replay memory that could worsen the off-policy gap.  
Something to do with ray Actor state corruption or something ? any clue?

**Comment by ericl** [comment]
Hm interesting, this never happens with 1 env per worker?

One thing that can happen with vectorized environments is excess correlation in experiences, since.they are all produced by the same policy. It might fix it to increase the train batch size to say 2000 to compensate.

**Comment by ericl** [comment]
It might also be worth looking at how the value function is doing (the vf_explained_var stat).

**Comment by praveen-palanisamy** [comment]
Run 1 was with 1 env per worker. So, the issue exists with and without vectorized envs.

The explained variance in the baseline value model also drops:
![vf_explained_var](https://user-images.githubusercontent.com/4770482/45459086-b6b59980-b6c4-11e8-8cb2-07ec26fe6f9f.PNG)

**Comment by ericl** [comment]
Maybe there's something exploding in value then. It's also possible that some straggler Ray actor returns  extremely stale batch data that causes this (we don't check the freshness of batches -- this could be solved by returning a timestamp with each sample batch and rejecting anything that looks too old).

On general principle I would try increasing the batch size, and also annealing the learning rate, e.g.
```
        lr_schedule: [
            [0, 0.0005],
            [20000000, 0.000000000001],
        ]
```

**Comment by praveen-palanisamy** [comment]
Thanks.
I didn't find any evidence for stale batches from Actors so far. So, that's good. 

The learning rate annealing with double the batch size (100) seems to help (ran only 1 experiment though) with the stabilization.

So, it's just the Hyper-parameter tuning that is required. I am closing the issue.
May be it's worthwhile to update the hyper-parameters in the "tuned_examples" so that it is actually "tuned"/stable.
